File: Matrix/ValidSudoku.py
# LeetCode 36 - Valid Sudoku
# Time Complexity: O(1) | Space Complexity: O(1)
from typing import List


class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rows = [[False] * 9 for _ in range(9)]
        cols = [[False] * 9 for _ in range(9)]
        boxes = [[False] * 9 for _ in range(9)]

        for row in range(9):
            for col in range(9):
                cell = board[row][col]
                if cell == ".":
                    continue

                digit = ord(cell) - ord("1")
                box = (row // 3) * 3 + col // 3
                if rows[row][digit] or cols[col][digit] or boxes[box][digit]:
                    return False

                rows[row][digit] = True
                cols[col][digit] = True
                boxes[box][digit] = True

        return True

File: Matrix/ValidSudoku.java
// LeetCode 36 - Valid Sudoku
// Time Complexity: O(1) | Space Complexity: O(1)
public class ValidSudoku {
    public boolean isValidSudoku(char[][] board) {
        boolean[][] rows = new boolean[9][9];
        boolean[][] cols = new boolean[9][9];
        boolean[][] boxes = new boolean[9][9];

        for (int row = 0; row < 9; row++) {
            for (int col = 0; col < 9; col++) {
                char cell = board[row][col];
                if (cell == '.') {
                    continue;
                }

                int digit = cell - '1';
                int box = (row / 3) * 3 + col / 3;
                if (rows[row][digit] || cols[col][digit] || boxes[box][digit]) {
                    return false;
                }

                rows[row][digit] = true;
                cols[col][digit] = true;
                boxes[box][digit] = true;
            }
        }

        return true;
    }
}
